client/lib/parsing.rs

```diff
@@ -1014,266 +1035,141 @@ mod tests {
+                session_path => happy::PATH, conflict: session_package_name => happy::PACKAGE_NAME, requires[] path_conflicts_with_package_name,
+                session_path => happy::PATH, conflict: session_hash =>         happy::HASH,         requires[] path_conflicts_with_hash,
+                session_path => happy::PATH, conflict: session_name =>         happy::HASH,         requires[] path_conflicts_with_name,
+                session_path => happy::PATH, conflict: session_version =>      happy::VERSION,      requires[] path_conflicts_with_version,
+                session_path => happy::PATH, conflict: session_entry_point =>  happy::ENTRY_POINT,  requires[] path_conflicts_with_entry_point,
+
+                // name
+                session_name => happy::NAME, conflict: session_package_hash => happy::PACKAGE_HASH, requires[session_entry_point => happy::ENTRY_POINT] name_conflicts_with_package_hash,
+                session_name => happy::NAME, conflict: session_package_name => happy::PACKAGE_NAME, requires[session_entry_point => happy::ENTRY_POINT] name_conflicts_with_package_name,
+                session_name => happy::NAME, conflict: session_hash =>         happy::HASH,         requires[session_entry_point => happy::ENTRY_POINT] name_conflicts_with_hash,
+
+                // hash
+                session_hash => happy::HASH, conflict: session_package_hash => happy::PACKAGE_HASH, requires[session_entry_point => happy::ENTRY_POINT] hash_conflicts_with_package_hash,
+                session_hash => happy::HASH, conflict: session_package_name => happy::PACKAGE_NAME, requires[session_entry_point => happy::ENTRY_POINT] hash_conflicts_with_package_name,
+                // name <-> hash is already checked
+                // name <-> path is already checked
+
+                // package_name
+                session_package_name => happy::PACKAGE_NAME, conflict: session_package_hash => happy::PACKAGE_HASH, requires[session_entry_point => happy::ENTRY_POINT] package_name_conflicts_with_package_hash
+                // package_name <-> hash is already checked
+                // package_name <-> name is already checked
+                // package_name <-> path is already checked
+
+                // package_hash
+                // package_name <-> package_hash is already checked
+                // package_name <-> hash is already checked
+                // package_name <-> name is already checked
+                // package_name <-> path is already checked
+
+            ]
+        ];
+
+        impl_test_matrix![
+            type: PaymentStrParams,
+            context: "parse_payment_info",
+            module: payment_str_params,
+            tests[
+
+                // amount
+                payment_amount => happy::PATH, conflict: payment_package_hash => happy::PACKAGE_HASH, requires[] amount_conflicts_with_package_hash,
```

Comment:
This all really needs to be well documented.  I'd like to see at least the result of `cargo expand` for one of these as a comment.

Reply:
`cargo expand` is _extremely_ verbose here, due to the internal calls to other macros like `format!` and `#[test]`. I've inspected the results of `expand` and ensured accuracy, but hand-unrolled the macro for the example.